compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -36,9 +66,249 @@ impl<'cx, 'a> Context<'cx, 'a> {
+                            DUMMY_NODE_ID,
+                        ),
+                    ]),
+                    span: self.span,
+                }),
+            ),
+        )
+    }
+
+    /// The necessary custom `panic!(...)` expression.
+    ///
+    /// panic!(
+    ///     "Assertion failed: ... \n With expansion: ...",
+    ///     __capture0,
+    ///     ...
+    /// );
+    fn build_panic(&self, expr_str: &str, fmt_str: &str, panic_path: Path) -> P<Expr> {
+        let escaped_expr_str = escape_to_fmt(expr_str);
+        let initial = [
+            TokenTree::token(
+                token::Literal(token::Lit {
+                    kind: token::LitKind::Str,
+                    symbol: Symbol::intern(&if fmt_str.is_empty() {
+                        format!("Assertion failed: {escaped_expr_str}")
+                    } else {
+                        format!("Assertion failed: {escaped_expr_str}\nWith captures:\n{fmt_str}")
+                    }),
+                    suffix: None,
+                }),
+                self.span,
+            ),
+            TokenTree::token(token::Comma, self.span),
+        ];
+        let captures = self.capture_decls.iter().flat_map(|cap| {
+            [
+                TokenTree::token(token::Ident(cap.ident.name, false), cap.ident.span),
+                TokenTree::token(token::Comma, self.span),
+            ]
+        });
+        self.cx.expr(
+            self.span,
+            ExprKind::MacCall(MacCall {
+                path: panic_path,
+                args: P(MacArgs::Delimited(
+                    DelimSpan::from_single(self.span),
+                    MacDelimiter::Parenthesis,
+                    initial.into_iter().chain(captures).collect::<TokenStream>(),
+                )),
+                prior_type_ascription: None,
+            }),
+        )
+    }
+
+    /// Recursive function called until `cond_expr` and `fmt_str` are fully modified.
+    ///
+    /// See [Self::manage_initial_capture] and [Self::manage_try_capture]
+    fn manage_cond_expr(&mut self, expr: &mut P<Expr>, fmt_str: &mut String) {
+        match (*expr).kind {
+            ExprKind::Binary(_, ref mut lhs, ref mut rhs) => {
+                self.manage_cond_expr(lhs, fmt_str);
+                self.manage_cond_expr(rhs, fmt_str);
+            }
+            ExprKind::Path(_, ref path) => {
+                let string = pprust::path_to_string(&path);
+                self.manage_initial_capture(expr, string, fmt_str);
+            }
+            _ => {}
+        }
+    }
+
+    /// Pushes the top-level declarations and modifies `expr` to try capturing variables.
+    ///
+    /// `fmt_str`, the formatting string used for debugging, is constructed to show possible
+    /// captured variables.
+    fn manage_initial_capture(
+        &mut self,
+        expr: &mut P<Expr>,
+        expr_string: String,
+        fmt_str: &mut String,
+    ) {
+        if self.paths.contains(&expr_string) {
```

Comment:
I don't like this string comparison scheme. I worry it will break down if someone uses a macro that internally creates variables that technically collide with variables from callers. Not sure how to improve this beyond going all in with comparing the paths themselves. Are we interested in anything beyond plain variable names here at all? I guess we can capture constants by path, too...

I expect this to get very messy with generics in the paths and such. I'd prefer to stick with very simple paths (single entry) for now and just put the `Ident` into the `paths` set.

Since I'm assuming more complex examples will actually include field projections and similar, this will change in follow up PRs, but that's alright. Even (or especially) then I think we should keep a structured representation around instead of doing rendered string matching.

Reply:
OK, changed from `String` to `Ident` considering only paths with a single segment.